verilog/analysis/checkers/instance_shadow_rule.cc

```diff
@@ -0,0 +1,140 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/instance_shadow_rule.h"
+
+#include <set>
+#include <string>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/bound_symbol_manager.h"
+#include "common/analysis/matcher/core_matchers.h"
+#include "common/analysis/matcher/matcher.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/CST/identifier.h"
+#include "verilog/CST/verilog_matchers.h"  // IWYU pragma: keep
+#include "verilog/CST/verilog_nonterminals.h"
+#include "verilog/analysis/descriptions.h"
+#include "verilog/analysis/lint_rule_registry.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+using verible::LintRuleStatus;
+using verible::LintViolation;
+using verible::SyntaxTreeContext;
+using verible::matcher::Matcher;
+
+// Register InstanceShadowRule
+VERILOG_REGISTER_LINT_RULE(InstanceShadowRule);
+
+absl::string_view InstanceShadowRule::Name() { return "instance-shadowing"; }
+const char InstanceShadowRule::kTopic[] = "mark-shadowed-instances";
+const char InstanceShadowRule::kMessage[] =
+    "Instance shadows the already declared variable";
+
+std::string InstanceShadowRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat(
+      "Checks that any defined variable does not shadow other variable in the "
+      "same scope",
+      GetStyleGuideCitation(kTopic), ".");
+}
+
+static const Matcher& InstanceShadowMatcher() {
+  static const Matcher matcher(SymbolIdentifierLeaf());
+  return matcher;
+}
+static bool isInAllowedNode(const SyntaxTreeContext& ctx) {
+  return ctx.IsInside(NodeEnum::kSeqBlock) ||
+         ctx.IsInside(NodeEnum::kGenvarDeclaration) ||
+         ctx.IsInside(NodeEnum::kReference);
+}
+
+void InstanceShadowRule::HandleSymbol(const verible::Symbol& symbol,
+                                      const SyntaxTreeContext& context) {
+  verible::matcher::BoundSymbolManager manager;
+  bool omit_node = false;
+  if (!InstanceShadowMatcher().Matches(symbol, &manager)) {
+    return;
+  }
+  const auto& labels = FindAllSymbolIdentifierLeafs(symbol);
+  if (labels.empty()) return;
+  // if the considered symbol name is not a declaration
+  if (context.IsInside(NodeEnum::kReference)) return;
+
+  const auto& rcontext = reversed_view(context);
+  const auto& rdirectParent = *std::next(rcontext.begin());
+  // we are looking for the potential labels that might overlap the considered
+  // declaration. We are searching all the labels within the visible scope
+  // until we find the node or we reach the top of the scope
+  for (auto rc = rcontext.begin(); rc != rcontext.end(); rc++) {
```

Comment:
Might be interesting to measure what performance impact going through nested loops for each symbol has in practice on a larger project.